### PR TITLE
Fill: Priority locks when placing and does not swap.

### PR DESCRIPTION
**Pull request by Berserker66**
Please format your title with what portion of the project this pull request is
targeting and what it's changing.

ex. "MyGame4: implement new game" or "Docs: add new guide for customizing MyGame3"

## What is this fixing or adding?
Should fix the generator spending very large amounts of time cramming an item into a priority location that will never fit.

## How was this tested?
About to be against the async gen

### CODE REVIEW

**Comment by Alchav** [comment]
Locking before accessibility corrections means no accessibility corrections on priority locations, which could lead to failures again

**Comment by Alchav** [comment]
But really I don't think locking while placing makes any difference to the change to not swap items. And it definitely makes the `accessibility_corrections` that is run right after priority fill useless

**Comment by Berserker66** [comment]
*should* be fixed now, and also faster than it used to be using a new on_place hook instead of scanning of copying all locations and scanning over them.

**Review by Alchav — APPROVED** [review]

**Comment by black-sliver** [comment]
If someone could come up with a bunch of yamls that should be affected by this, and see if failure rate changes, that'd be nice.

i.e. `ap-roller.py "https://github.com/ArchipelagoMW/Archipelago:7f3f886e41653e2df2c6f56500672146cc05c2db" "#1099" -O --repeat 1 --slots 6 --seeds 1-20 --include my_yaml_dir` or something

**Comment by Berserker66** [comment]
I don't think it affects failure rate. Priority likes to go into swapping mode, often fail, then quit swapping mode. Then it gets resolved into logic fill. As a result, we just waste a whole lot of time but don't fail the generation.

File: Fill.py [hunk 1]
```diff
@@ -23,7 +23,8 @@ def sweep_from_pool(base_state: CollectionState, itempool: typing.Sequence[Item]
 
 
 def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations: typing.List[Location],
-                     itempool: typing.List[Item], single_player_placement: bool = False, lock: bool = False) -> None:
+                     itempool: typing.List[Item], single_player_placement: bool = False, lock: bool = False,
+                     swap: bool = True, on_place: typing.Callable[[Location], None] = None) -> None:
```
Comment (black-sliver) [inline Fill.py:27]: mypy does not complain, but shouldn't that be Optional[Callable] if it defaults to None?
```suggestion
                     swap: bool = True, on_place: typing.Optional[typing.Callable[[Location], None]] = None) -> None:
```

File: Fill.py [hunk 2]
```diff
@@ -272,19 +278,26 @@ def distribute_items_restrictive(world: MultiWorld) -> None:
...
-    fill_restrictive(world, world.state, prioritylocations, progitempool)
+    # "priority fill"
+    fill_restrictive(world, world.state, prioritylocations, progitempool, swap=False, on_place=mark_for_locking)
     accessibility_corrections(world, world.state, prioritylocations, progitempool)
 
-    for location in prioritylocations_lock:
-        if location.item:
-            location.locked = True
+    for location in lock_later:
+        location.locked = True
```
Comment (black-sliver) [inline Fill.py:293]: i may be wrong, but to me it looks like accessibility_corrections can remove items from priority locations? so we have to check if location has an item.
```suggestion
        if location.item:
            location.locked = True
```

File: Fill.py (hunk 1 above)
Comment (Berserker66) [reply Fill.py:27]: Prior to 3.10 docs it was written that this Optional is optional, but current python docs state that the optional Optional is now desired for explicitness sake, so yes. I guess mypy still allows 3.9 and earlier conformity

File: Fill.py (hunk 2 above)
Comment (Berserker66) [reply Fill.py:293]: corrections keeps track of all locations that were emptied and then runs another fill_restrictive on them,
so it's save to assume they're filled

**Review by black-sliver — APPROVED** [review]

**Pull request merged**